Convert net-shell to use new shell

jukkar opened the pull request:
Fixes #8875

jukkar committed:
net: shell: Migrate to new shell API

Convert net-shell to use the new shell API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jarz-nordic commented on subsys/net/ip/net_shell.c:
```diff
@@ -3534,159 +3753,249 @@ static void iface_vlan_cb(struct net_if *iface, void *user_data)
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_nbr)
+{
+	SHELL_CMD(rm, NULL, "Remove neighbor from cache.",
```
You are defining help string for this command (and others):
```
SHELL_CMD(rm, NULL, "**Remove neighbor from cache.**",
		  cmd_net_nbr_rm),
```
However in the command handler: **cmd_net_nbr_rm** you are not calling a function resposible for printing help.
In current command implementation when user will call command: 
`net nbr rm -h` or `net nbr rm --help`
help string "Remove neighbor from cache." will not be dispayed by shell.

To be able to see help string you need to add following code at the beginning of command halder:

`if (shell_help_requested(shell)) {
		shell_help_print(shell, opt, opt_len);
		return;
}`

nordic-krch replied:
In that case I would recommend using shell_cmd_precheck() function:
> Prints help if request and prints error message on wrong argument count.

jarz-nordic replied:
Hint: Function `shell_help_print` prints help for existing command and subcommands 1 level below.

I prefere to add this function to each command handler, for either root command or subcommands.

nordic-krch commented on subsys/net/ip/net_shell.c:
```diff
@@ -3534,159 +3753,249 @@ static void iface_vlan_cb(struct net_if *iface, void *user_data)
+	SHELL_CMD(rpl, NULL, "Show RPL mesh routing status.", cmd_net_rpl),
+	SHELL_CMD(stacks, NULL, "Show network stacks information.",
+		  cmd_net_stacks),
+	SHELL_CMD(stats, &net_cmd_stats, "Show network statistics.",
```
You could solve 'stats' command in more handy way by utilizing dynamic autocompletion. It could have syntax like:
- _net stats_ prints stats for all interfaces
- _net_stats <iface1> <iface2>_ - prints stats for all named interfaces
Very similar approach is taken in _log disable_ command (https://github.com/zephyrproject-rtos/zephyr/blob/f2596952599df87d9dafaa8d692f7f737a0fb3f6/subsys/logging/log_cmds.c#L390)

jukkar replied:
That would be cool, there are also other commands that could benefit from this.

jukkar replied:
I have been experimenting with the dynamic commands. I added dynamic completion for `net iface up <idx>` command. I got it working but it is not very user friendly atm.
In this example I have two network interfaces. When completing the command I see this:
```
net:~$ net iface up <tab>
  0  1
```
but it is not easy to understand what the values actually mean. For that one needs to add --help string in which case I get
```
net:~$ net iface up 1 --help
up - 'net iface up <iface index>' takes network interface up.
Options:
  -h, --help  :Show command help.
Subcommands:
  0  :Ethernet (0x00421860)
  1  :Ethernet (0x00421920)
```

For user this is quite cumbersome. Why not show the help during completion like this:
```
net:~$ net iface up <tab>
  0  :Ethernet (0x00421860)
  1  :Ethernet (0x00421920)
```

This would be much more user friendly and user could select the proper index immediately. If there is only one interface, then it is selected automatically currently and this is just fine.

jarz-nordic replied:
So you would like to have help string for promted dynamic commands or all commands?
Or maybe I could prepare special type of dynamic commands printing this help?

jukkar replied:
I was just setting `entry->help  = set_iface_index_help(idx);` for dynamic command, and was expecting that it's value would be used in completion. Will send new version shortly so you can see how I did this.

jarz-nordic replied:
Setting `entry->help` only allows to print help when command is called with -h or --help.
Otherwise it would be inconsistency between static and dynamic commands.

When you press tab you only see command syntax, not command syntax + help.
Maybe you could change a dynamic command syntax name from `0` to `0 :Ethernet (0x00421860)` ?

jukkar replied:
> Maybe you could change a dynamic command syntax name from 0 to 0 :Ethernet (0x00421860) ?

I tried that too as that seemed like a logical thing to do. So after pressing <tab>, the "0 :Ethernet (0x00421860)" and "1: Ethernet <(0xabcd01234)" was printed, and by selecting 0 and then pressing <tab> again the "0 :Ethernet (0x00421860)" is selected which is of course wrong. Naturally one would probably not press <tab> after selecting 0 but the point is that the expansion goes wrong in this case.
So I was just wondering if the "entry->help" string could be used here to guide user what the options means when pressing <tab>

jarz-nordic replied:
I am not quite sure I understand you. Please correct me if I am wrong:
1. You have changed dynamic command syntax from `0` to `0 :Ethernet (0x00421860)` and from `1` to `1: Ethernet <(0xabcd01234)`
2. When you press Tab button you can see prompted 2 options: `0 :Ethernet (0x00421860)` and "1: Ethernet <(0xabcd01234)".
3. You are pressing `0` next Tab and shell has autocomplete `0` to  `0 :Ethernet (0x00421860)` and you did not expect it?

jarz-nordic replied:
To be honest I do not see easy solution for now to prompt command syntax + command help but to autocompleate only command name. :(

jukkar replied:
I will probably remove the dynamic completion support from net-shell as the feature uses lot of memory. Especially the IPv6 neighbor completion uses 240 bytes memory by default. The network interface index completion is using smaller amount of memory but is currently not very user friendly.

jukkar replied:
> Could we use interface names, like eth0, at shell level?

Currently we do not have a concept for interface name in IP stack, if we introduce such a name it would be only for shell which complicates things. This needs to be pondered more, atm there are more important things to do so this is left for later (at least from my part). Patches are welcome of course.

> if we make this option configurable.

Yes, I was planning to make the completion configurable and enable it by default.

nordic-krch commented on subsys/net/ip/net_shell.c:
```diff
@@ -3534,159 +3753,249 @@ static void iface_vlan_cb(struct net_if *iface, void *user_data)
+
+SHELL_CMD_REGISTER(net, &net_commands, "Networking commands", NULL);
+
+SHELL_UART_DEFINE(shell_transport_uart);
```
you should leave shell definition to the application. Here only commands should be specified.

jukkar replied:
Hmm, with net-shell the idea is that the app only needs to set CONFIG_NET_SHELL=y and get a shell automatically.

nordic-krch replied:
note, that in future there might be multiple transports (net too) and even multiple shell instances available so that approach will not fly for long. We might approach that the other way: add UART as default transport to the shell module.

jukkar replied:
@nordic-krch The `SHELL_DEFINE()` macro is a bit problematic as it requires the transport as a parameter. As a user I would just like to set `CONFIG_NET_SHELL=y` and expecting that net-shell is enabled without any configuration. Could the transport parameter left out or set to NULL? The shell subsystem would then just select the default backend in that case. Setting the backend here looks something that would be better done in some other part of the system.

nordic-krch commented on subsys/net/ip/net_shell.c:
```diff
@@ -3534,159 +3753,249 @@ static void iface_vlan_cb(struct net_if *iface, void *user_data)
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_nbr)
+{
+	SHELL_CMD(rm, NULL, "Remove neighbor from cache.",
```
if it is possible you could try to provide dynamic subcommand to be able to use tab to help pick a neighbor.

nordic-krch commented:
@jukkar, i'm very happy to see this. I've left few comments in the review.

jukkar commented on subsys/net/ip/net_shell.c:
```diff
@@ -199,21 +211,25 @@ static struct ethernet_capabilities eth_hw_caps[] = {
 	for (i = 0; i < ARRAY_SIZE(eth_hw_caps); i++) {
 		if (caps & eth_hw_caps[i].capability) {
-			printk("\t%s\n", eth_hw_caps[i].description);
+			PR("\t%s\n", eth_hw_caps[i].description);
```
Tried to keep the macro name short, the lines are quite long already. I originally had the full `shell_fprintf(shell, SHELL_NORMAL, ....);` but that was just way too long.

nordic-krch replied:
ok

jukkar commented on subsys/net/ip/net_shell.c:
```diff
@@ -271,12 +289,12 @@ static void iface_cb(struct net_if *iface, void *user_data)
+			PR("\t%d: Qav ", i);
 			if (ret) {
-				printk("not supported\n");
+				PR("not supported\n");
```
Ok, makes sense.

jukkar commented:
Added dynamic command support for "net iface" and "net nbr" commands. Also added color output if there is error/warning/info message printed.

jukkar commented:
Fixed merge conflict.

jukkar commented:
Rebased against jarz-nordic:shell_rework

jarz-nordic commented:
@jukkar : Currently @nordic-krch is working on this branch to solve all conflicts. I hope it will be ready later today.

jukkar commented:
> Currently @nordic-krch is working on this branch to solve all conflicts. I hope it will be ready later today.

Ok, no worries. The code just bit rots easily if there is new stuff merged into master. I wish we could merge the #9362 quite soon in order to avoid this constant rebasing.

jukkar commented:
It seems that this PR needs to go in before the networking logging changes at #8816 as legacy-shell does not work properly with new logging.

jukkar commented:
The issue with net logging was fixed already, proceeding with assumption that network logging pr at #8816 is merged first, and this after that.

jukkar committed:
net: wifi: Convert wifi shell to use new shell

Use new shell instead of the legacy one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ieee802154: Convert IEEE 802.15.4 shell to use new shell

Use new shell instead of the legacy one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka reviewed (approved):
seems fine to me

jukkar commented:
Resolved merge conflict

jukkar committed:
samples: net: rpl_border_router: Convert to use new shell

Use new shell instead of the legacy one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: wpanusb: Removing useless shell

There were no shell commands implemented so remove the shell
support from this sample application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: zperf: Convert to use the new shell

Use new shell instead of the legacy one. This commit also fixes
the configuration file mess and allows the program to be run on
more hardware.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: bt: Convert network Bluetooth shell to use new shell

Use new shell instead of the legacy one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: http_server: Remove Quark devboard from sanity

ROM consumption is higher with new shell which prevents
quark_se_c1000_devboard from building. Remove that board
from sanitychecker runs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed various sanitycheck errors. This needs still more testing and TLC before can be merged. For example if one enables more than one shell modules, the history stops working. This can be seen in zperf which enables net-shell and has also its own shell.

jukkar committed:
net: openthread: Convert to use new shell

Use new shell instead of the legacy one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Converting OpenThread shell to new shell.

jukkar commented:
Fixed sanity

jukkar commented:
Updated because of changes in shell APIs.

jarz-nordic commented:
@jukkar : What about `*.conf` files under this location: tests/bluetooth/shell ?
Maybe not with this PR but I guess the need to be updated

jukkar commented:
> What about `*.conf` files under this location: tests/bluetooth/shell ?

I hope @Vudentz will look into these. Are you otherwise ok with the PR?

jarz-nordic commented:
@Vudentz : Please update `*.conf` files under: tests/bluetooth/shell to get rid off old shell dependencies.

jukkar commented:
Fixed merge conflict.

@nordic-krch does this look now ok for you?

jarz-nordic commented on subsys/net/ip/Kconfig:
```diff
@@ -35,11 +35,26 @@ source "subsys/net/ip/Kconfig.ipv4"
 	  Activate shell module that provides network commands like
 	  ping to the console.
 
+config NET_SHELL_DYN_CMD_COMPLETION
```
Is this implemented / needed ?

jukkar replied:
The reason for this setting is that if one wishes to autocomplete IPv6 addresses for example, then we need to allocate some space for these ip addresses. As each IPv6 address string can take up to 46 bytes of memory and we might have for example 8 of them defined, that would occupy over 300 bytes of memory. I made this configurable because of that so that if the device is limited with memory, then this dynamic cmd support can be turned off.
